<br>

# Milestone1_1. Data Collection

## Team #30

In [1]:
import pandas as pd
import json
import requests

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.


<br>

## 1.1. Load & Check Original Yelp Academic Data

Source: https://www.yelp.com/dataset/download

Google drive: https://drive.google.com/drive/u/0/folders/1XYDq_uKP3-SZCPCvuyfTZzs8v_yvULj3

In [2]:
# Import yelp business data
# It takes quite some time

business_data = []
with open('data/yelp_academic_dataset_business.json') as f:
    for line in f:
        business_data.append(json.loads(line))

#importing yelp checkin data
checkin_data = []
with open('data/yelp_academic_dataset_checkin.json') as f:
    for line in f:
        checkin_data.append(json.loads(line))

#importing yelp review data
review_data = []
with open('data/yelp_academic_dataset_review.json') as f:
    for line in f:
        review_data.append(json.loads(line))

#importing yelp tip data
tip_data = []
with open('data/yelp_academic_dataset_tip.json') as f:
    for line in f:
        tip_data.append(json.loads(line))

#importing yelp user data
user_data = []
with open('data/yelp_academic_dataset_user.json') as f:
    for line in f:
        user_data.append(json.loads(line))

#creating business dataframe from json format
business_df = pd.DataFrame(business_data)

#creating checkin dataframe from json format
checkin_df = pd.DataFrame(checkin_data)

#creating review dataframe from json format
review_df = pd.DataFrame(review_data)

#creating tip dataframe from json format
tip_df = pd.DataFrame(tip_data)

#creating user dataframe from json format
user_df = pd.DataFrame(user_data)

In [3]:
# Check business dataframe - columns, shape, and unique business_id
print(business_df.columns)
print(f'data shape: {business_df.shape}')
print(f'unique business ids: {business_df.business_id.nunique()}')
business_df.head(2)

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'attributes', 'categories', 'hours'],
      dtype='object')
data shape: (150346, 14)
unique business ids: 150346


business_id                      name  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   

                           address           city state postal_code  \
0           1616 Chapala St, Ste 2  Santa Barbara    CA       93101   
1  87 Grasso Plaza Shopping Center         Affton    MO       63123   

    latitude   longitude  stars  review_count  is_open  \
0  34.426679 -119.711197    5.0             7        0   
1  38.551126  -90.335695    3.0            15        1   

                               attributes  \
0           {'ByAppointmentOnly': 'True'}   
1  {'BusinessAcceptsCreditCards': 'True'}   

                                          categories  \
0  Doctors, Traditional Chinese Medicine, Naturop...   
1  Shipping Centers, Local Services, Notaries, Ma...   

                                               hours  
0                                               None  
1  {'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', ...

In [4]:
# Check non-US locations
business_df.state.unique()

array(['CA', 'MO', 'AZ', 'PA', 'TN', 'FL', 'IN', 'LA', 'AB', 'NV', 'ID',
       'DE', 'IL', 'NJ', 'NC', 'CO', 'WA', 'HI', 'UT', 'TX', 'MT', 'MI',
       'SD', 'XMS', 'MA', 'VI', 'VT'], dtype=object)

In [5]:
business_df[business_df.state=='XMS'].business_id.nunique()

1

In [6]:
# Alberta, Canada
business_df[business_df.state=='AB'].business_id.nunique()

5573

In [7]:
# d1 = {'ALABAMA' : 'AL', 'ALASKA' : 'AK', 'ARIZONA' : 'AZ', 'ARKANSAS' : 'AR', 'CALIFORNIA' : 'CA',
#       'COLORADO' : 'CO', 'CONNECTICUT' : 'CT', 'DELAWARE' : 'DE', 'FLORIDA' : 'FL', 'GEORGIA' : 'GA',
#       'HAWAII' : 'HI', 'IDAHO' : 'ID', 'ILLINOIS' : 'IL', 'INDIANA' : 'IN', 'IOWA' : 'IA', 'KANSAS' : 'KS',
#       'KENTUCKY' : 'KY', 'LOUISIANA' : 'LA', 'MAINE' : 'ME', 'MARYLAND' : 'MD', 'MASSACHUSETTS' : 'MA',
#       'MICHIGAN' : 'MI', 'MINNESOTA' : 'MN', 'MISSISSIPPI' : 'MS', 'MISSOURI' : 'MO', 'MONTANA' : 'MT',
#       'NEBRASKA' : 'NE', 'NEVADA' : 'NV', 'NEW HAMPSHIRE' : 'NH', 'NEW JERSEY' : 'NJ', 'NEW MEXICO' : 'NM',
#       'NEW YORK' : 'NY', 'NORTH CAROLINA' : 'NC', 'NORTH DAKOTA' : 'ND', 'OHIO' : 'OH', 'OKLAHOMA' : 'OK',
#       'OREGON' : 'OR', 'PENNSYLVANIA' : 'PA', 'RHODE ISLAND' : 'RI', 'SOUTH CAROLINA' : 'SC',
#       'SOUTH DAKOTA' : 'SD', 'TENNESSEE' : 'TN', 'TEXAS' : 'TX', 'UTAH' : 'UT', 'VERMONT' : 'VT',
#       'VIRGINIA' : 'VA', 'WASHINGTON' : 'WA', 'WEST VIRGINIA' : 'WV', 'WISCONSIN' : 'WI', 'WYOMING' : 'WY',
#       'PUERTO RICO' : 'PR', 'DISTRICT OF COLUMBIA' : 'DC'}

In [8]:
# Check checkin dataframe - columns, shape, and unique business_id
print(checkin_df.columns)
print(f'data shape: {checkin_df.shape}')
print(f'unique business ids: {checkin_df.business_id.nunique()}')
checkin_df.head(2)

Index(['business_id', 'date'], dtype='object')
data shape: (131930, 2)
unique business ids: 131930


business_id                                               date
0  ---kPU91CF4Lq2-WlRu9Lw  2020-03-13 21:10:56, 2020-06-02 22:18:06, 2020...
1  --0iUa4sNDFiZFrAdIWhZQ  2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011...

In [9]:
# Check review dataframe - columns, shape, and unique business_id
print(review_df.columns)
print(f'data shape: {review_df.shape}')
print(f'unique business ids: {review_df.business_id.nunique()}')
review_df.head(2)

Index(['review_id', 'user_id', 'business_id', 'stars', 'useful', 'funny',
       'cool', 'text', 'date'],
      dtype='object')
data shape: (6990280, 9)
unique business ids: 150346


review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   

   stars  useful  funny  cool  \
0    3.0       0      0     0   
1    5.0       1      0     1   

                                                text                 date  
0  If you decide to eat here, just be aware it is...  2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year...  2012-01-03 15:28:18

In [10]:
# Check tip dataframe - columns, shape, and unique business_id
print(tip_df.columns)
print(f'data shape: {tip_df.shape}')
print(f'unique business ids: {tip_df.business_id.nunique()}')
tip_df.head(2)

Index(['user_id', 'business_id', 'text', 'date', 'compliment_count'], dtype='object')
data shape: (908915, 5)
unique business ids: 106193


user_id             business_id  \
0  AGNUgVwnZUey3gcPCJ76iw  3uLgwr0qeCNMjKenHJwPGQ   
1  NBN4MgHP9D3cw--SnauTkA  QoezRbYQncpRqyrLH6Iqjg   

                                                text                 date  \
0                     Avengers time with the ladies.  2012-05-18 02:17:21   
1  They have lots of good deserts and tasty cuban...  2013-02-05 18:35:10   

   compliment_count  
0                 0  
1                 0

In [11]:
# Check user dataframe - columns, shape, and unique user_id
print(user_df.columns)
print(f'data shape: {user_df.shape}')
print(f'unique business ids: {user_df.user_id.nunique()}')
user_df.head(2)

Index(['user_id', 'name', 'review_count', 'yelping_since', 'useful', 'funny',
       'cool', 'elite', 'friends', 'fans', 'average_stars', 'compliment_hot',
       'compliment_more', 'compliment_profile', 'compliment_cute',
       'compliment_list', 'compliment_note', 'compliment_plain',
       'compliment_cool', 'compliment_funny', 'compliment_writer',
       'compliment_photos'],
      dtype='object')
data shape: (1987897, 22)
unique business ids: 1987897


user_id    name  review_count        yelping_since  useful  \
0  qVc8ODYU5SZjKXVBgXdI7w  Walker           585  2007-01-25 16:47:26    7217   
1  j14WgRoU_-2ZE1aw1dXrJg  Daniel          4333  2009-01-25 04:35:42   43091   

   funny   cool                                              elite  \
0   1259   5994                                               2007   
1  13066  27281  2009,2010,2011,2012,2013,2014,2015,2016,2017,2...   

                                             friends  fans  ...  \
0  NSCy54eWehBJyZdG2iE84w, pe42u7DcCH2QmI81NX-8qA...   267  ...   
1  ueRPE0CX75ePGMqOFVj6IQ, 52oH4DrRvzzl8wh5UXyU0A...  3138  ...   

   compliment_more  compliment_profile  compliment_cute  compliment_list  \
0               65                  55               56               18   
1              264                 184              157              251   

   compliment_note  compliment_plain  compliment_cool  compliment_funny  \
0              232               844              467               467   
1             1847              7054             3131              3131   

   compliment_writer  compliment_photos  
0                239                180  
1               1521               1946  

[2 rows x 22 columns]

<br>

## 1.2. Subset Yelp Data

In [12]:
# Check unique restaurant in the U.S.
business_df[business_df.categories.str.contains('restaurant', case = False, na = False)].business_id.nunique()

52286

In [13]:
# Check unique restaurant in CA
business_df[business_df.categories.str.contains('restaurant', case = False, na = False)
            & (business_df['state'] == 'CA')].business_id.nunique()

1161

In [14]:
# Unique business_id
restaurant_id = business_df[business_df.categories.str.contains('restaurant', case = False, na = False)
                           & ~(business_df['state'].isin(['AB','XMS']))].business_id.unique()

ca_restaurant_id = business_df[business_df.categories.str.contains('restaurant', case = False, na = False)
                           & (business_df['state'] == 'CA')].business_id.unique()

In [15]:
# Unique business_id in subset of review
review_df[review_df.business_id.isin(ca_restaurant_id)].business_id.nunique()

1161

In [16]:
# Unique business_id in subset of tip
tip_df[tip_df.business_id.isin(ca_restaurant_id)].business_id.nunique()

1077

In [17]:
# Subset of business table
restaurant_df = business_df[business_df.business_id.isin(restaurant_id)].copy()
ca_restaurant_df = business_df[business_df.business_id.isin(ca_restaurant_id)].copy()

In [18]:
# Subset of review table
restaurant_review_df = review_df[review_df.business_id.isin(restaurant_id)].copy()
ca_restaurant_review_df = review_df[review_df.business_id.isin(ca_restaurant_id)].copy()

In [19]:
# Subset of tip table
restaurant_tip_df = tip_df[tip_df.business_id.isin(restaurant_id)].copy()
ca_restaurant_tip_df = tip_df[tip_df.business_id.isin(ca_restaurant_id)].copy()

In [20]:
# Subset of user table
restaurant_review_user_id = restaurant_review_df.user_id.unique()
ca_restaurant_review_user_id = ca_restaurant_review_df.user_id.unique()

restaurant_user_df = user_df[user_df.user_id.isin(restaurant_review_user_id)].copy()
ca_restaurant_user_df = user_df[user_df.user_id.isin(ca_restaurant_review_user_id)].copy()

<br>

## 1.3. Check Attributes

In [21]:
restaurant_df.loc[3,'attributes']

{'RestaurantsDelivery': 'False',
 'OutdoorSeating': 'False',
 'BusinessAcceptsCreditCards': 'False',
 'BusinessParking': "{'garage': False, 'street': True, 'validated': False, 'lot': False, 'valet': False}",
 'BikeParking': 'True',
 'RestaurantsPriceRange2': '1',
 'RestaurantsTakeOut': 'True',
 'ByAppointmentOnly': 'False',
 'WiFi': "u'free'",
 'Alcohol': "u'none'",
 'Caters': 'True'}

In [22]:
restaurant_df.loc[85,'attributes']

{'RestaurantsTakeOut': 'True',
 'NoiseLevel': "u'average'",
 'Caters': 'True',
 'Ambience': "{'touristy': False, 'hipster': True, 'romantic': False, 'divey': False, 'intimate': False, 'trendy': True, 'upscale': False, 'classy': False, 'casual': True}",
 'RestaurantsReservations': 'False',
 'BusinessAcceptsCreditCards': 'True',
 'RestaurantsTableService': 'False',
 'GoodForKids': 'True',
 'RestaurantsPriceRange2': '2',
 'WheelchairAccessible': 'True',
 'OutdoorSeating': 'True',
 'RestaurantsDelivery': 'None',
 'HasTV': 'False',
 'RestaurantsAttire': "'casual'",
 'Alcohol': "u'none'",
 'GoodForMeal': "{'dessert': False, 'latenight': False, 'lunch': True, 'dinner': False, 'brunch': True, 'breakfast': True}",
 'DogsAllowed': 'True',
 'RestaurantsGoodForGroups': 'True',
 'HappyHour': 'False',
 'BusinessParking': "{'garage': False, 'street': True, 'validated': False, 'lot': False, 'valet': False}",
 'BikeParking': 'True',
 'WiFi': "u'no'"}

In [23]:
restaurant_df.loc[85,'categories']

'Food, Restaurants, Salad, Coffee & Tea, Breakfast & Brunch, Sandwiches, Bakeries'

In [24]:
review_df.loc[0,'text']

"If you decide to eat here, just be aware it is going to take about 2 hours from beginning to end. We have tried it multiple times, because I want to like it! I have been to it's other locations in NJ and never had a bad experience. \n\nThe food is good, but it takes a very long time to come out. The waitstaff is very young, but usually pleasant. We have just had too many experiences where we spent way too long waiting. We usually opt for another diner or restaurant on the weekends, in order to be done quicker."

In [25]:
tip_df.loc[1,'text']

'They have lots of good deserts and tasty cuban sandwiches'

In [26]:
tip_df.loc[1,'compliment_count']

0

In [27]:
review_df.columns

Index(['review_id', 'user_id', 'business_id', 'stars', 'useful', 'funny',
       'cool', 'text', 'date'],
      dtype='object')

In [28]:
tip_df.compliment_count.describe()

count    908915.000000
mean          0.012525
std           0.120763
min           0.000000
25%           0.000000
50%           0.000000
75%           0.000000
max           6.000000
Name: compliment_count, dtype: float64

In [29]:
checkin_df.iloc[1,1]

'2010-09-13 21:43:09, 2011-05-04 23:08:15, 2011-07-18 22:30:31, 2012-09-07 20:28:50, 2013-03-27 15:57:36, 2013-08-13 00:31:34, 2013-08-13 00:31:48, 2013-09-23 17:39:38, 2013-11-18 06:34:08, 2014-04-12 23:04:47'

<br>

## 1.4. Flatten Columns

In [30]:
# Flatten the 'attributes' column
flattened_attributes = pd.json_normalize(restaurant_df['attributes'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'attributes' column
restaurant_df.drop(columns=['attributes'], inplace=True)


In [31]:
# Flatten the 'BusinessParking' column
flattened_attributes = pd.json_normalize(restaurant_df['BusinessParking'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'BusinessParking' column
restaurant_df.drop(columns=['BusinessParking'], inplace=True)

In [32]:
# Flatten the 'Ambience' column
flattened_attributes = pd.json_normalize(restaurant_df['Ambience'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'Ambience' column
restaurant_df.drop(columns=['Ambience'], inplace=True)

In [33]:
# Flatten the 'GoodForMeal' column
flattened_attributes = pd.json_normalize(restaurant_df['GoodForMeal'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'GoodForMeal' column
restaurant_df.drop(columns=['GoodForMeal'], inplace=True)

In [34]:
# Flatten the 'Music' column
flattened_attributes = pd.json_normalize(restaurant_df['Music'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'Music' column
restaurant_df.drop(columns=['Music'], inplace=True)

In [35]:
# Flatten the 'BestNights' column
flattened_attributes = pd.json_normalize(restaurant_df['BestNights'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'BestNights' column
restaurant_df.drop(columns=['BestNights'], inplace=True)

In [36]:
# Flatten the 'DietaryRestrictions' column
flattened_attributes = pd.json_normalize(restaurant_df['DietaryRestrictions'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'DietaryRestrictions' column
restaurant_df.drop(columns=['DietaryRestrictions'], inplace=True)

In [37]:
# Flatten the 'HairSpecializesIn' column
flattened_attributes = pd.json_normalize(restaurant_df['HairSpecializesIn'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'HairSpecializesIn' column
restaurant_df.drop(columns=['HairSpecializesIn'], inplace=True)

In [38]:
# Flatten the 'hours' column
flattened_attributes = pd.json_normalize(restaurant_df['hours'].apply(lambda x: {} if pd.isna(x) else x))

# Concatenate the flattened attributes with the original DataFrame
restaurant_df = pd.concat([restaurant_df.reset_index(drop=True),
                           flattened_attributes.reset_index(drop=True)], axis=1)

# Drop the original 'hours' column
restaurant_df.drop(columns=['hours'], inplace=True)

In [39]:
ca_restaurant_df = restaurant_df[restaurant_df.business_id.isin(ca_restaurant_id)].copy()

<br>

## 1.5. Store Data for Subsequent Use

In [40]:
restaurant_df.to_csv('data/restaurant.csv',index=False)
ca_restaurant_df.to_csv('data/ca_restaurant.csv',index=False)

restaurant_review_df.to_csv('data/review.csv',index=False)
ca_restaurant_review_df.to_csv('data/ca_review.csv',index=False)

restaurant_tip_df.to_csv('data/tip.csv',index=False)
ca_restaurant_tip_df.to_csv('data/ca_tip.csv',index=False)

restaurant_user_df.to_csv('data/user.csv',index=False)
ca_restaurant_user_df.to_csv('data/ca_user.csv',index=False)

<br>

## 1.6. Demographic Data from US Census API

In [41]:
# Extract zip code from merged dataframe
print(f'entries missing zip code: {restaurant_df.postal_code.isna().sum()}')
print(f'entries missing ca zip code: {ca_restaurant_df.postal_code.isna().sum()}')

zip_codes = restaurant_df['postal_code'].unique()
ca_zip_codes = ca_restaurant_df['postal_code'].unique()

print(f'number of unique zip codes: {len(zip_codes)}')
print(f'number of unique CA zip codes: {len(ca_zip_codes)}')

entries missing zip code: 0
entries missing ca zip code: 0
number of unique zip codes: 898
number of unique CA zip codes: 16


In [42]:
# Connecting to US Census's API to get demographic data for each zip code
# https://www.census.gov/data/developers/data-sets/acs-5year.html

def get_census_data(zip_code, api_key):
    base_url = "https://api.census.gov/data"
    year = "2022"
    dataset = "acs/acs5"

    variables = {
        "B19013_001E": "median_household_income",
        "B01003_001E": "population",
        "B11016_001E": "household_cnt",
        "B15003_001E": "education_total",
        "B15003_022E": "bachelors_degree",
        "B01002_001E": "median_age",
        "B03001_003E": "population_hispanic_latino",
        "B02001_002E": "population_white", 
        "B02001_005E": "population_asian",
    }

    # optional attributes: density; hh_income_greaterThan100k_pcnt;

    params = {
        "get": ",".join(variables.keys()),
        "for": f"zip code tabulation area:{zip_code}",
        "key": api_key
    }

    response = requests.get(f"{base_url}/{year}/{dataset}", params=params)
    if response.status_code == 200:
        data = response.json()[1]
        result = {variables[key]: value for key, value in zip(variables.keys(), data)}
        result["zip_code"] = zip_code
        return result
    else:
        return f"Error: {response.status_code}"

api_key = "bbde78e5f070664a3d06156001d4c374fc6c7737"
zip_code = "94085"
census_data = get_census_data(zip_code, api_key)
print(census_data)

{'median_household_income': '168250', 'population': '22996', 'household_cnt': '8952', 'education_total': '17152', 'bachelors_degree': '3581', 'median_age': '32.9', 'population_hispanic_latino': '6891', 'population_white': '5269', 'population_asian': '10692', 'zip_code': '94085'}


### Document of demographic variables

### * median_household_income:
This is the median income of households within the ZIP code area. The value '168250' indicates that the median household income is $168,250 per year. The median is the middle value when all household incomes are arranged in order, so half of the households earn more than this amount and half earn less.

### * population:
This indicates the total number of people living in the ZIP code area. The value '22996' means that there are 22,996 individuals residing in this area.

### * household_cnt:
This number represents the count of households in the ZIP code area, not the size of each household. The value '8952' means there are 8,952 households. To find the average household size, you would typically divide the population by the number of households.

### * education_total:
This is the total number of people in the ZIP code area who are above a certain age (usually 25) and have some level of educational attainment. The value '17152' is the count of such individuals. This variable is often used as a denominator to calculate education-related percentages.

### * bachelors_degree:
This number indicates the count of individuals who have attained a bachelor's degree. The value '3581' means that 3,581 people in this area have a bachelor's degree. You can use this number along with the education_total to find the percentage of the population with a bachelor's degree.

### * median_age:
This is the median age of all residents in the ZIP code area. The value '32.9' means that half of the population is younger than 32.9 years and half is older. This is a measure of the age distribution of the population.


### * population_hispanic_latino:
This indicates the total number of Hispanic or Latino people living in the ZIP code area. 


### * population_white:
This indicates the total number of Caucasian people living in the ZIP code area. 


### * population_asian:
This indicates the total number of Asian people living in the ZIP code area. 

#### <font color=red> below cell takes a while to run (~1 hour)</font>

In [43]:
# Retreiving API data
# It takes quite some time

api_key = "bbde78e5f070664a3d06156001d4c374fc6c7737"

data_list = []
for zip_code in zip_codes:
    data = get_census_data(zip_code, api_key)
    if isinstance(data, dict):
        data_list.append(data)
    else:
        print(f"Failed to retrieve data for ZIP code {zip_code}: {data}")

demographic_df = pd.DataFrame(data_list)

print(demographic_df.shape)

Failed to retrieve data for ZIP code 08018: Error: 204
Failed to retrieve data for ZIP code 37243: Error: 204
Failed to retrieve data for ZIP code 89555: Error: 204
Failed to retrieve data for ZIP code : Error: 400
Failed to retrieve data for ZIP code 19176: Error: 204
Failed to retrieve data for ZIP code 93118: Error: 204
Failed to retrieve data for ZIP code 19357: Error: 204
Failed to retrieve data for ZIP code 89595: Error: 204
Failed to retrieve data for ZIP code 19443: Error: 204
Failed to retrieve data for ZIP code 18931: Error: 204
Failed to retrieve data for ZIP code 70063: Error: 204
Failed to retrieve data for ZIP code 70016: Error: 204
Failed to retrieve data for ZIP code 19409: Error: 204
Failed to retrieve data for ZIP code 19395: Error: 204
Failed to retrieve data for ZIP code 83701: Error: 204
Failed to retrieve data for ZIP code 33588: Error: 204
Failed to retrieve data for ZIP code 70170: Error: 204
Failed to retrieve data for ZIP code 90261: Error: 204
Failed to retri

In [44]:
demographic_df.head()

median_household_income population household_cnt education_total  \
0                   67690      15232          8612           11555   
1                   70271      19104          7347           13799   
2                   71054      50347         22083           36811   
3                   59973      40144         16034           27024   
4                   88878      17631          9754           14004   

  bachelors_degree median_age population_hispanic_latino population_white  \
0             4148       31.1                        943             8110   
1             1873       40.9                       1191            17780   
2             7927       40.7                       1924            43306   
3             5699       34.8                       3914            11592   
4             5130       36.3                       4003            11664   

  population_asian zip_code  
0             3723    19107  
1               30    37015  
2             2151    63123  
3              298    37207  
4              696    33602

In [45]:
#store demographic data for subsequent use
demographic_df.to_csv('data/demographics.csv',index=False)